README.md
# Pently

A music engine for NES that supports both NES playback (with sound
effects) and NSF.

Documentation can be found in `docs/` and example tracks can be found in `musicseq.pently`.

## Building

### Prerequisites

- Python 3
- [ca65](https://cc65.github.io/cc65/)
- GNU Make and Coreutils - on Windows, install MSYS through
  [mingw-get](http://www.mingw.org/wiki/Getting_Started)
- For NES (not NSF) format: Pillow (Python imaging library)  
  UNIX: `python3 -m pip install pillow`  
  Windows: `py -3 -m pip install pillow`
- For FamiTracker conversion:
  [j0CC-FamiTracker](https://github.com/jimbo1qaz/j0CC-FamiTracker)
  and [ft2pently](https://github.com/NovaSquirrel/ft2pently)

For help setting up Python, ca65, Make, and Coreutils, see the README
file for [nrom-template](https://github.com/pinobatch/nrom-template).

### Building

Once you have the above installed, run `make` to build and play
or `make pently.nes` to only build a ROM.
Then edit the score and run `make` again to hear the changes.
To use an entirely different score file, open `makefile` and change
`scorename`, or use e.g. `make NTS.nsfe` or `make NTS.nes` to use
score file `audio/NTS.pently`.
To use a FamiTracker module, edit the `FAMITRACKER` and `FT2P` paths
in `makefile` to reflect executable paths on your system, then run
`make Foothills.nsf` to use score file `audio/Foothills.ftm`.

## License

Copyright © 2009-2018 Damian Yerrick.
Pently is free software, under the zlib License.


CHANGES.txt
0.05wip12 (future)
* Add port to ASM6 via a translation script (#9, requested by Pokun,
  Doug Fraker, and Joe Granato)
* pentlybss.py: Sort arpeggio interval variables
* pentlybss.py: ASM6-compatible output mode
* pentlyas.py: ASM6-compatible output mode

0.05wip11 (2019-03-14)
* Attack track allows grace notes (#35, reported by retrodpc)
* Channel volume in attack track is ignored rather than clobbering
  another variable (#35, reported by retrodpc)
* Option to round row grid to nearest tick on loop, pattern start,
  or beat (#37)
* Remove use of other .proc's labels in preparation for ASM6 port
* pentlyas.py: Fix bugs in fallthrough command and document it (#1,
  reported by jroatch)
* pentlyas.py: Specify tempo based on a musical duration other than a
  beat (#34)
* pentlyas.py: Calculate duration of each song and sound effect (#40)
* NES shell: Take track titles from score (#38)
* NSF shell: Take NSF title, author, and copyright from score (#38)
* NSF shell: Write NSFe and NSF2 formats (#39, #41)
* sfxed: Fix display of rate in slow (rate 10+) effects
* Move .pently files to audio folder
* Document workarounds for lack of reentrancy (#2)
* makefile: Add a recipe to convert a FamiTracker module with
  ft2pently (#36)

0.05wip10 (2018-10-28)
* Grace note durations no longer reduced by 1 frame
* Always write period high byte for triangle, fixing problem with
  zero volume in instruments (#32, reported by NovaSquirrel)
* Suppress linker warnings for export/import address size mismatch
* Overhaul names and scope of internal subroutines
* Rearrange conductor processing based on bytecode order
* Move sustain phase processing above pitch effects
* Remove redundant checks for attack injection in pitch effects
* Fix misbehavior when music and a slowed-down sound effect are
  played at once
* C bindings for commonly used functions (requested by adrian09_01)
* pentlyseq.inc: don't truncate patdef at 7 bits (reported and fixed
  by NovaSquirrel)
* pentlyas.py: a score can include other files, such as an instrument
  set (#29)
* pentlyas.py: o0 through o7 to set reference pitch
  (requested by zzo38)
* pentlyas.py: allow grace notes in drum patterns
* pentlyas.py: playing a pattern with only rests or waits is nonfatal
* pentlyas.py: fix warnings for missing "fine" or "dal segno"
* pentlyas.py: order of objects in the output is explicit, the same
  order in which they appear in the score (requested by NovaSquirrel)
* pentlyas.py: error if no songs are defined (reported by
  NovaSquirrel)
* pentlyas.py: -A means --period-tuning; --output means -o
* mkrammap.py: rename to pentlybss.py to encourage remembering
  to include it when vendoring Pently
* Profiler: don't overreport CPU use if the shell has a lag frame
* Document how to skip the makefile's default autoplay (#30)
* makefile: Allow building from an arbitrarily named score by
  giving the name of the corresponding .nes or .nsf
* Add pseudocode to aid in porting to non-6502 platforms
* Add Travis CI build script
* Sample songs: Include the soundtrack of all games by Damian Yerrick
  that have appeared in Action 53.  These are Concentration Room,
  Thwaite, Zap Ruder, and RHDE: Furniture Fight.

0.05wip9 (2018-02-21)
* PENTLY_USE option to disable music engine entirely
* PENTLY_USE option to disable 303 portamento
* PENTLY_USE option to enable triangle channel duty fix
* Regression fix: Cancel injection when attack ends (had been broken
  since slow arps)
* Cancel grace note when pattern begins
* Add rehearsal transport control to skip to a row or slow down or
  pause playback (#27)
* Add variable mix to mute tracks
* Store current pitch and volume for use by visualizer
* pentlyas.py: add warning infrastructure
* pentlyas.py: v1 through v4 as synonyms for pp through ff
* pentlyas.py: p (pause) is synonym for r (rest)
* pentlyas.py: q repeats last chord, as in LilyPond
* pentlyas.py: bar check in pattern to warn if not on a measure
  boundary (#26)
* pentlyas.py: pickup works in patterns as well
* pentlyas.py: unreachable A in tunings below A=437 is nonfatal
* pentlyas.py: mmloctaves off, which future chord syntax will require
* pentlyas.py: set rehearsal marks for navigation in score
* pentlyas.py: set initial playback position and track muting
* NES shell: add per-song screen with visualizer, rehearsal mark
  navigation, track muting, tempo, and pause-and-step
* Script to estimate pitch of 93-step noise
* Docs: Recalculate PENTLY_USE savings
* Import Sound Effects Editor as of 2014-09-10

0.05wip8 (2017-12)
* Reduce BSS (main RAM) footprint when several effects are off
* Reduce ZP footprint when attack phase is off
* Compress 0-pitch frames of instrument envelopes
* pentlyas.py: correctly parse drum names starting with "l"
* pentlyas.py: option to change overall tuning up or down from A440
* pentlyas.py: special filename - means standard output
* pentlyas.py: output lists each song's total size in comments
* Add "Isometry" song demonstrating attack track more thoroughly
* makefile: change score filename
* NES shell: move copyright notice to status bar
* NES shell: status shows ROM size and current and peak CPU usage
* NES/NSF executables no longer in Git; will be in releases instead

0.05wip7 (2017-09)
* Change to zlib license, which preserves the spirit of the
  original license but is recognizable by GitHub
* Add portamento, the single most requested feature
* paldetect: Interleave with vwait so as not to depend on NMI handler
* pentlyas.py: Add 8 predefined chord names
  (M, maj7, dom7, aug, m, m7, dim, dim7)
* pentlyas.py: Score can define new chord names
* pentlyas.py: Add single-note arpeggio syntax similar to LilyPond
  chord mode (thanks to retrodpc for testing)
* pentlyas.py: Option to name chord by its top note instead of bottom
* pentlyas.py: First and second inversions of chords
* Arpeggio can be set to slow or fast (requested by w7n)
* Add "Arpeggio Waltz" song demonstrating new features

src/vis.s
;
; Pently audio engine
; Visualizer
;
; Copyright 2018 Damian Yerrick
; 
; This software is provided 'as-is', without any express or implied
; warranty.  In no event will the authors be held liable for any damages
; arising from the use of this software.
; 
; Permission is granted to anyone to use this software for any purpose,
; including commercial applications, and to alter it and redistribute it
; freely, subject to the following restrictions:
; 
; 1. The origin of this software must not be misrepresented; you must not
;    claim that you wrote the original software. If you use this software
;    in a product, an acknowledgment in the product documentation would be
;    appreciated but is not required.
; 2. Altered source versions must be plainly marked as such, and must not be
;    misrepresented as being the original software.
; 3. This notice may not be removed or altered from any source distribution.
;

.include "pentlyconfig.inc"
.include "pently.inc"
.include "nes.inc"
.include "shell.inc"

ARROW_SPRITE_TILE = $15

; This screen covers visualization, rehearsal mark seeking,
; slowdown/step playback, and track muting.
;
; Local TODO for https://github.com/pinobatch/pently/issues/27
; 1. Tempo scaling
; 2. Playback stepping a row at a time
; 3. Check issue
; 4. What is causing a higher than expected Peak on song start?

DIRTY_MUTE = $01

.zeropage
vis_to_clear: .res 1
vis_dirty: .res 1
vis_section_load_row: .res 1
vis_A_gesture_time: .res 1
vis_cursor_x: .res 1

.if PENTLY_USE_REHEARSAL
  vis_cur_song_section: .res 1
  vis_num_sections: .res 1
  vis_section_src: .res 2
  vis_start_held: .res 1
.endif

; Injection palette
.if PENTLY_USE_VIS
  vis_pulse1_color: .res 1
  vis_pulse2_color: .res 1
  vis_tri_color: .res 1
  vis_noise_color: .res 1
  vis_mute: .res 1
.endif

CH_TRI = $08
CH_NOISE = $0C
CH_END = $10
.if PENTLY_USE_ATTACK_TRACK
  LAST_TRACK = $10
.else
  LAST_TRACK = $0C
.endif

copydst_lo = $0180
copydst_hi = $0181
copybuf = $0100
LF = $0A

songrow = 35
caporow = 36
muterow = 56

.code
.proc vis
  .if ::PENTLY_USE_REHEARSAL
    lda #caporow
    sta vis_section_load_row
    lda #2
    sta vis_to_clear
  .else
    lda #$FF
    sta vis_section_load_row
    lda #0
    sta vis_to_clear
  .endif

  jsr load_song_title

  lda #0
  .if ::PENTLY_USE_REHEARSAL
    sta vis_start_held
  .endif
  sta vis_A_gesture_time
  sta vis_cursor_x
  lda #DIRTY_MUTE
  sta vis_dirty
  ldx #4
  stx oam_used
vis_loop:
  ldx oam_used
  jsr ppu_clear_oam

  ; Wait for vblank
  lda nmis
  :
    cmp nmis
    beq :-

  lda vis_to_clear
  beq vis_cleared
    jsr vis_clear_part
    lda #VBLANK_NMI|OBJ_1000|BG_0000|0
    bne have_which_nt
  vis_cleared:
    lda #>OAM
    sta OAM_DMA
    .if ::PENTLY_USE_VIS
      ; Update palette
      lda #$3F
      sta PPUADDR
      lda #$11
      sta PPUADDR
      lda vis_pulse1_color
      ldx vis_noise_color
      sta PPUDATA
      stx PPUDATA
      bit PPUDATA
      bit PPUDATA
      lda vis_pulse2_color
      sta PPUDATA
      bit PPUDATA
      bit PPUDATA
      bit PPUDATA
      lda vis_tri_color
      sta PPUDATA
    .endif
    
    lda copydst_hi
    bmi nocopy
      sta PPUADDR
      lda copydst_lo
      sta PPUADDR
      ldx #0
      :
        lda copybuf,x
        sta PPUDATA
        inx
        cpx #32
        bcc :-
    
    nocopy:
    lda #VBLANK_NMI|OBJ_1000|BG_0000|1
    sta copydst_hi
  have_which_nt:

  ldx #0
  ldy #0
  sec
  jsr ppu_screen_on
  jsr pently_update
  jsr read_pads
  ldx #4
  stx oam_used
  jsr vis_prepare_vram_row
  .if ::PENTLY_USE_VIS
    jsr vis_update_obj
  .endif
  .if ::PENTLY_USE_VARMIX
    jsr vis_draw_mute_cursor
  .endif

  lda vis_A_gesture_time
  beq :+
    dec vis_A_gesture_time
  :
  
  .if ::PENTLY_USE_REHEARSAL
    jsr find_current_row

    lda vis_cursor_x
    bne notRehearsalKeys
      jsr vis_draw_section_arrow
      jsr vis_handle_rehearsal_keys
    notRehearsalKeys:
    jsr vis_handle_tempo_keys
  .endif
  
  .if ::PENTLY_USE_VARMIX
    jsr vis_handle_varmix_keys
  .endif

  lda new_keys
  and #KEY_B
  bne vis_done
  jmp vis_loop
vis_done:
  rts
.endproc

; Song name and section names ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

.proc clear_copybuf
  sec
  ror a
  sta copydst_hi
  lda #0
  ror a
  lsr copydst_hi
  ror a
  lsr copydst_hi
  ror a
  sta copydst_lo
  ldx #31
  lda #' '
  :
    sta copybuf,x
    dex
    bpl :-
  bail:
  rts
.endproc

.proc load_song_title
bail = clear_copybuf::bail
src = $00
newlines_left = $03
  lda #songrow
  jsr clear_copybuf
  lda #>tracknames_txt
  ldy #<tracknames_txt
  sta src+1

  lda cur_song
  beq found
  sta newlines_left
  lda #0
  sta src+0
  searchloop:
    lda (src),y
    beq bail
    iny
    bne :+
      inc src+1
    :
    cmp #LF
    bne searchloop
    dec newlines_left
    bne searchloop
  found:
  sty src+0
  ldx #2
.endproc
.proc print_to_copybuf
src = $00
  ldy #0
  copyloop:
    lda (src),y
    beq not_found
    cmp #LF
    beq not_found
    sta copybuf,x
    iny
    inx
    cpx #32
    bcc copyloop
  not_found:
  rts
.endproc

.proc vis_clear_part
  ; Clearing nametable must be FAST
  clc
  adc #$23
  sta PPUADDR
  lda #$80
  sta PPUADDR
  lda #' '
  ldx #64
  :
    sta PPUDATA
    sta PPUDATA
    sta PPUDATA
    sta PPUDATA
    dex
    bne :-
  dec vis_to_clear
bail:
  rts
.endproc

.proc vis_prepare_vram_row
no_room = vis_clear_part::bail

  ; Ensure there's room in the VRAM copy buffer to load something
  lda copydst_hi
  bpl no_room

  ; Now look for things to load
  .if ::PENTLY_USE_REHEARSAL
    ; Are the section names drawn yet?
    lda vis_section_load_row
    bmi not_section_name
      jmp vis_draw_section_name
    not_section_name:
  .endif

  .if ::PENTLY_USE_VARMIX
    lda vis_dirty
    and #DIRTY_MUTE
    beq not_dirty0
      jmp vis_draw_mute
    not_dirty0:
  .endif

  rts
.endproc

.if PENTLY_USE_REHEARSAL
.proc vis_draw_section_name
src = $00

  lda vis_section_load_row
  jsr clear_copybuf
  lda vis_section_load_row
  sec
  sbc #caporow + 1
  bcs past_caporow
    ; 'Capo' row: Load the word "capo"
    lda #'c'
    sta copybuf+3
    lda #'a'
    sta copybuf+4
    lda #'p'
    sta copybuf+5
    lda #'o'
    sta copybuf+6

    ; With that out of the way, we can find the rehearsal marks
    lda cur_song
    asl a
    tax
    lda pently_rehearsal_marks,x
    sta vis_section_src+0
    lda pently_rehearsal_marks+1,x
    sta vis_section_src+1
    ldy #0
    lda (vis_section_src),y
    sta $5555
    sta vis_num_sections
    beq no_marks_left

    ; Skip past the row counts to the mark names
    asl a
    adc #2
    adc vis_section_src
    sta vis_section_src
    bcc :+
      inc vis_section_src+1
    :
    inc vis_section_load_row
    rts
  past_caporow:
    lda vis_section_src
    sta src
    lda vis_section_src+1
    sta src+1
    ldx #3
    jsr print_to_copybuf
    
    ; If at NUL terminator, stop
    cmp #0
    beq no_marks_left

    ; Move to next piece
    sec
    tya
    adc vis_section_src
    sta vis_section_src
    bcc :+
      inc vis_section_src+1
    :
    inc vis_section_load_row
    rts
no_marks_left:

  lda #$FF
  sta vis_section_load_row
  rts
.endproc

;;
; Sets vis_cur_song_section to the lowest section number whose ending
; row number is greater than the current row number.
.proc find_current_row
src = $00
nleft = $02

  ; If rows not printed yet, assume section 0
  lda vis_section_load_row
  cmp #caporow+16
  lda #0
  bcc have_section

  lda cur_song
  asl a
  tax
  lda pently_rehearsal_marks,x
  sta src+0
  lda pently_rehearsal_marks+1,x
  sta src+1
  ldy #0
  lda (src),y
  beq have_section
  sta nleft
  iny
  iny
  sectloop:
    lda pently_rowslo
    cmp (src),y
    iny
    lda pently_rowshi
    sbc (src),y
    bcc have_y
    iny
    dec nleft
    bne sectloop
  have_y:
  tya
  lsr a
  sec
  sbc #1
have_section:
  sta vis_cur_song_section
  rts
.endproc

.proc vis_draw_section_arrow
  ldy oam_used
  tya
  clc
  adc #4
  sta oam_used
  lda vis_cur_song_section
  asl a
  asl a
  asl a
  adc #(caporow - 32) * 8 - 1
  sta OAM,y
  lda #RIGHT_ARROW_SPRITE_TILE
  sta OAM+1,y
  lda #0
  sta OAM+2,y
  lda #16
  sta OAM+3,y
bail:
  rts
.endproc

.proc seek_to_section
src = $00
  asl a
  beq vis_draw_section_arrow::bail
  tay
  lda cur_song
  asl a
  tax
  lda pently_rehearsal_marks,x
  sta src+0
  lda pently_rehearsal_marks+1,x
  sta src+1
  iny
  lda (src),y
  tax
  dey
  lda (src),y
  jmp pently_skip_to_row
.endproc

.proc vis_handle_rehearsal_keys
  lda cur_keys
  and #KEY_START
  bne startHeld

  ; Down: Go to start of next section
  lda new_keys
  and #KEY_DOWN
  beq notDown
  lda vis_cur_song_section
  cmp vis_num_sections
  bcs notDown
    lda #1
    adc vis_cur_song_section
    jsr seek_to_section
  notDown:

  ; Up: Rewind, then go to start of previous section
  lda new_keys
  and #KEY_UP
  beq notUp
  lda vis_cur_song_section
  beq notUp
    lda pently_tempo_scale
    pha
    lda cur_song
    jsr pently_start_music
    .if ::PENTLY_USE_VARMIX
      jsr vis_set_mute
    .endif
    pla
    sta pently_tempo_scale
    lda vis_cur_song_section
    sec
    sbc #1
    bcc notUp
    jsr seek_to_section
  notUp:

startHeld:
  rts
.endproc

.proc vis_handle_tempo_keys
  lda new_keys
  and #KEY_START
  ora vis_start_held
  sta vis_start_held
  lda cur_keys
  eor #$FF
  and vis_start_held
  beq not_start_release
  lda #$80
  eor pently_tempo_scale
  sta pently_tempo_scale
  jmp consume_start_held

not_start_release:
  ; Test for Start+Up and Start+Down
  lda cur_keys
  and #KEY_START
  beq nope

  lda pently_tempo_scale
  and #$03
  tax

  lda new_keys
  and #KEY_UP
  beq not_start_up
    cpx #0
    beq :+
      dec pently_tempo_scale
    :
    jmp consume_start_held
  not_start_up:
  
  lda new_keys
  and #KEY_DOWN
  beq nope
    cpx #3
    beq :+
      inc pently_tempo_scale
    :

consume_start_held:
  lda #0
  sta vis_start_held
nope:

  ; Check for A while paused = step once
  
  lda pently_tempo_scale
  and new_keys
  bpl not_paused_A
  lda vis_cursor_x
  bne not_paused_A
    clc
    ldx pently_rowshi
    lda pently_rowslo
    adc #1
    bcc :+
      inx
    :
    jsr pently_skip_to_row
  not_paused_A:

  rts
.endproc

.endif

; Visualizer ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

VIS_PITCH_Y = 183
TARGET_NOTE_DOT_TILE = $13

.if PENTLY_USE_VIS

.proc vis_update_obj
semitonenum = $08
overtone_x = $09
effective_vol = $0A
semitone_xlo = $0C
semitone_xhi = $0D

  ldy #4

  ; Y: OAM index; X: channel
  ldx #0
  chloop:
    ; Draw pitch/volume marks
    lda pently_vis_dutyvol,x
    and #$0F
    beq pitchvol_mark_done  ; Draw no such mark for silent channels
      jmp draw_pitchvol_mark
    pitchvol_mark_done:

    ; Draw target note marks for pulse 1, pulse 2 and triangle
    cpx #CH_NOISE
    bcs target_note_mark_done
      jmp draw_target_note_mark
    target_note_mark_done:
    inx
    inx
    inx
    inx
    cpx #CH_END
    bcc chloop
  sty oam_used

  ; Calculate palette based on attack injection
.if ::PENTLY_USE_ATTACK_TRACK
  ldx #$20
  lda pently_vis_arpphase+0
  bmi pulse1_injected
.endif
    ldx #$26
  pulse1_injected:
  stx vis_pulse1_color

.if ::PENTLY_USE_ATTACK_TRACK
  ldx #$20
  lda pently_vis_arpphase+4
  bmi pulse2_injected
.endif
    ldx #$2A
  pulse2_injected:
  stx vis_pulse2_color

.if ::PENTLY_USE_ATTACK_TRACK
  ldx #$20
  lda pently_vis_arpphase+CH_TRI
  bmi tri_injected
.endif
    ldx #$12
  tri_injected:
  stx vis_tri_color

  ; Calculate noise palette based on color
  ldx #$10
  lda pently_vis_pitchhi+CH_NOISE
  bpl noise_is_hiss
    ldx #$28
  noise_is_hiss:
  stx vis_noise_color
  rts
    
draw_pitchvol_mark:
  lsr a
  cpx #CH_TRI
  bne :+
    lda #0
  :
  sta effective_vol

  lda pently_vis_pitchhi,x
  cpx #CH_NOISE
  bne st_not_noise
    stx semitone_xlo
    and #$0F
    tax
    lda noise_to_sprite_x,x
    ldx semitone_xlo
    jmp have_semitone_xhi
  st_not_noise:

  ; Triangle draws the lowest octave as hollow and the other
  ; octaves as solid, except offset an octave to the left because
  ; it sounds an octave below pulse.
  cpx #CH_TRI
  bne have_semitonenum
    sta semitonenum
    cmp #12  ; C=0 for hollow (lowest) octave, 1 for solid octaves
    lda #3 * 12 * 5
    bcs :+
      lda #3 * 12 * 4
      inc effective_vol
    :
    sta overtone_x
    lda semitonenum
    bcc :+
      sbc #12
    :
  have_semitonenum:
  sta semitonenum
  sta semitone_xhi
  lda pently_vis_pitchlo,x
  asl a
  rol semitone_xhi
  sta semitone_xlo
  clc
  adc pently_vis_pitchlo,x
  bpl :+
    inc semitone_xhi
  :
  lda semitonenum
  adc semitone_xhi
  clc
  adc #16
have_semitone_xhi:
  sta semitone_xhi

  lda #VIS_PITCH_Y  ; Y position
  sta OAM+0,y
  lda effective_vol
  ora vis_ch_vol_tiles,x
  sta OAM+1,y
  lda vis_ch_attribute,x
  sta OAM+2,y
  lda semitone_xhi
  sta OAM+3,y
  iny
  iny
  iny
  iny

  cpx #CH_TRI
  bne no_overtone_mark
  clc
  adc overtone_x
  bcs no_overtone_mark
    sta OAM+3,y
    lda #VIS_PITCH_Y
    sta OAM+0,y
    lda #$12  ; triangle overtone mark
    sta OAM+1,y
    lda #$02  ; triangle attribute
    sta OAM+2,y
    iny
    iny
    iny
    iny
  no_overtone_mark:
  jmp pitchvol_mark_done

draw_target_note_mark:
  ; First, is this a black key or a white key?
  lda pently_vis_note,x
  cmp #96
  bcc :+
    sbc #96
  :
  cmp #48
  bcc :+
    sbc #48
  :
  cmp #24
  bcc :+
    sbc #24
  :
  cmp #12
  bcc :+
    sbc #12
  :
  
  stx semitone_xlo
  tax
  lda whitekey_pos,x
  lsr a
  ldx semitone_xlo
  sta semitone_xhi
  
  lda vis_ch_whitekey_y,x
  bcc :+
    lda vis_ch_blackkey_y,x
  :
  sta OAM+0,y
  lda #TARGET_NOTE_DOT_TILE
  sta OAM+1,y
  lda vis_ch_attribute,x
  sta OAM+2,y
  
  ; Find the X position of the mark
  lda pently_vis_note,x
  cpx #CH_TRI
  bne target_not_trioctave
  cmp #12
  bcc target_not_trioctave
    sbc #12
  target_not_trioctave:
  sta semitonenum
  asl a
  clc
  adc semitonenum
  clc
  adc semitone_xhi
  sta OAM+3,y
  iny
  iny
  iny
  iny
  jmp target_note_mark_done
.endproc

.rodata

vis_ch_whitekey_y:
  ; 0. Y base for white keys
  ; 1. Y position for black keys
  ; 2. Tile base (for volume)
  ; 3. attribute
  .byte 170, 157, $00, 0
  .byte 172, 161, $00, 1
  .byte 174, 165, $10, 2
  .byte 255, 255, $08, 0
vis_ch_blackkey_y = vis_ch_whitekey_y + 1
vis_ch_vol_tiles = vis_ch_whitekey_y + 2
vis_ch_attribute = vis_ch_whitekey_y + 3

; Approximate X positions of noise pitches, assuming 93-step
; waveform.
noise_to_sprite_x:
  .byte 248, 212, 176, 140, 104, 83, 68, 57, 45, 33, 12
  ; The following are not to scale with the rest of the keyboard.
  ; If they were, they'd lie to the left of the keyboard's left edge.
  .byte 10, 8, 6, 3, 0

whitekey_pos:
  .byte 16*2+0
  .byte       16*2+1
  .byte 15*2+0
  .byte 17*2+0
  .byte       16*2+1
  .byte 16*2+0
  .byte       16*2+1
  .byte 15*2+0
  .byte 17*2+0
  .byte       16*2+1
  .byte 16*2+0
  .byte       16*2+1

.endif  ; PENTLY_USE_VIS

; Track muting ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
.code
.if PENTLY_USE_VARMIX
.proc vis_set_initial_mute
  lda #pently_resume_mute
  sta vis_mute
.endproc
.proc vis_set_mute
mutebit = $08
  lda vis_mute
  ldx #0
  loop:
    lsr a
    ror pently_mute_track,x
    inx
    inx
    inx
    inx
    cpx #LAST_TRACK+4
    bcc loop
  rts
.endproc

MUTE_X = 15
.proc vis_draw_mute
  lda #<~DIRTY_MUTE
  sta vis_dirty
  lda #muterow
  jsr clear_copybuf
  
  ldx #31
  :
    lda chmute_status_bar,x
    sta copybuf,x
    dex
    bpl :-

  ldx #15
  lda vis_mute
  chloop:
    lsr a
    bcc :+
      tay
      lda #$FE
      sta copybuf,x
      tya
    :
    cpx #(LAST_TRACK / 4) * 3 + MUTE_X
    inx
    inx
    inx
    bcc chloop
  rts
.endproc

.proc vis_handle_varmix_keys
MUTE_ALL = (1 << (LAST_TRACK / 4 + 1)) - 1
  lda vis_cursor_x
  beq notLeft
  lda new_keys
  and #KEY_LEFT
  beq notLeft
    dec vis_cursor_x
    jmp cancel_A_gesture
  notLeft:

  lda vis_cursor_x
  cmp #(LAST_TRACK / 4) + 1
  bcs notRight
  lda new_keys
  and #KEY_RIGHT
  beq notRight
    inc vis_cursor_x
  cancel_A_gesture:
    lda #0
    sta vis_A_gesture_time
  notA:
    rts
  notRight:

  ldx vis_cursor_x
  beq notA
  lda new_keys
  bpl notA
  lda vis_A_gesture_time
  bne is_double_click
    ; Single click: Toggle this channel
    lda #15
    sta vis_A_gesture_time
    lda one_shl_x-1,x
    eor vis_mute
    jmp have_vis_mute
    
is_double_click:
  ; Double-click with at least one other channel not muted puts this
  ; channel on solo
  lda vis_mute
  eor #MUTE_ALL  ; A: channels NOT muted
  and not_1_shl_x-1,x  ; A: channels other than this not muted
  beq unmute_all
    lda not_1_shl_x-1,x
    and #MUTE_ALL
    jmp have_vis_mute
  unmute_all:
    lda #0
  have_vis_mute:

  sta vis_mute
  lda #DIRTY_MUTE
  ora vis_dirty
  sta vis_dirty
  jmp vis_set_mute
.endproc

.proc vis_draw_mute_cursor
  lda vis_cursor_x
  beq no_cursor
    ldy oam_used
    tya
    clc
    adc #4
    sta oam_used
    
    lda #(muterow - 32) * 8 + 4
    sta OAM+0,y
    lda #ARROW_SPRITE_TILE
    sta OAM+1,y
    lda #1
    sta OAM+2,y
    lda vis_cursor_x
    asl a
    adc vis_cursor_x
    asl a
    asl a
    asl a
    adc #(MUTE_X - 3) * 8 + 4
    sta OAM+3,y
  no_cursor:
  rts
.endproc

.rodata
chmute_status_bar:
  .byte $02,$02,$08,$09,$0A,$0B,$0C,$0D,$0E,$0F,$02,$02,$02,$02,$02,$FF
  .byte $EA,$EB,$FF,$FA,$FB,$FF,$EC,$ED,$FF,$FC,$FD,$FF,$EE,$EF,$02,$02
one_shl_x:
  .repeat 8, I
    .byte 1 << I
  .endrepeat
not_1_shl_x:
  .repeat 8, I
    .byte $FF ^ (1 << I)
  .endrepeat

.endif
